src/sqlfluff/rules/L022.py

```diff
@@ -63,19 +63,20 @@ def _eval(self, context: RuleContext) -> Optional[List[LintResult]]:
             )
             for idx, seg in enumerate(expanded_segments):
                 if seg.is_type("bracketed"):
-                    # Check if the preceding keyword is AS, otherwise it's a column name
+                    # Check if the preceding segment is AS, otherwise it's a column name
                     # definition in the CTE.
-                    preceding_keyword = next(
+                    preceding_code = next(
```

Comment:
The variable name says "code", but we're stopping on `not whitespace`, which I think would also stop on comments. Is this the desired behavior? Maybe rename this variable `preceding_non_whitespace` (clunky but accurate)?

Reply:
Comments should be excluded too. Updated code and added test case.

Reply:
(I hope the above makes sense!) I thought about this before the first attempted fix, but I thought I could solve it without restructuring, but now I'm thinking it's proven tricky enough that it may be the right approach.

Reply:
I think that might indeed be possible, but would be a much bigger rewrite and beyond my skills - one you're better suited for?

So currently a sample query is this:

```sql
    WITH
    cte_1 AS (
        SELECT 1 AS var
    ),

    cte_2 (var) AS (
        SELECT 2
    )

    SELECT
        cte_1.var,
        cte_2.var
    FROM cte_1, cte_2;
```

This creates this parse tree:

```
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    whitespace:                                               '    '
[L:  1, P:  5]      |    statement:
[L:  1, P:  5]      |        with_compound_statement:
[L:  1, P:  5]      |            keyword:                                          'WITH'
[L:  1, P:  9]      |            newline:                                          '\n'
[L:  2, P:  1]      |            whitespace:                                       '    '
[L:  2, P:  5]      |            common_table_expression:
[L:  2, P:  5]      |                identifier:                                   'cte_1'
[L:  2, P: 10]      |                whitespace:                                   ' '
[L:  2, P: 11]      |                keyword:                                      'AS'
[L:  2, P: 13]      |                whitespace:                                   ' '
[L:  2, P: 14]      |                bracketed:
[L:  2, P: 14]      |                    start_bracket:                            '('
[L:  2, P: 15]      |                    [META] indent:
[L:  2, P: 15]      |                    newline:                                  '\n'
[L:  3, P:  1]      |                    whitespace:                               '        '
[L:  3, P:  9]      |                    select_statement:
[L:  3, P:  9]      |                        select_clause:
[L:  3, P:  9]      |                            keyword:                          'SELECT'
[L:  3, P: 15]      |                            [META] indent:
[L:  3, P: 15]      |                            whitespace:                       ' '
[L:  3, P: 16]      |                            select_clause_element:
[L:  3, P: 16]      |                                literal:                      '1'
[L:  3, P: 17]      |                                whitespace:                   ' '
[L:  3, P: 18]      |                                alias_expression:
[L:  3, P: 18]      |                                    keyword:                  'AS'
[L:  3, P: 20]      |                                    whitespace:               ' '
[L:  3, P: 21]      |                                    identifier:               'var'
[L:  3, P: 24]      |                        [META] dedent:
[L:  3, P: 24]      |                    newline:                                  '\n'
[L:  4, P:  1]      |                    whitespace:                               '    '
[L:  4, P:  5]      |                    [META] dedent:
[L:  4, P:  5]      |                    end_bracket:                              ')'
[L:  4, P:  6]      |            comma:                                            ','
[L:  4, P:  7]      |            newline:                                          '\n'
[L:  5, P:  1]      |            newline:                                          '\n'
[L:  6, P:  1]      |            whitespace:                                       '    '
[L:  6, P:  5]      |            common_table_expression:
[L:  6, P:  5]      |                identifier:                                   'cte_2'
[L:  6, P: 10]      |                whitespace:                                   ' '
[L:  6, P: 11]      |                bracketed:
[L:  6, P: 11]      |                    start_bracket:                            '('
[L:  6, P: 12]      |                    [META] indent:
[L:  6, P: 12]      |                    identifier_list:
[L:  6, P: 12]      |                        identifier:                           'var'
[L:  6, P: 15]      |                    [META] dedent:
[L:  6, P: 15]      |                    end_bracket:                              ')'
[L:  6, P: 16]      |                whitespace:                                   ' '
[L:  6, P: 17]      |                keyword:                                      'AS'
[L:  6, P: 19]      |                whitespace:                                   ' '
[L:  6, P: 20]      |                bracketed:
[L:  6, P: 20]      |                    start_bracket:                            '('
[L:  6, P: 21]      |                    [META] indent:
[L:  6, P: 21]      |                    newline:                                  '\n'
[L:  7, P:  1]      |                    whitespace:                               '        '
[L:  7, P:  9]      |                    select_statement:
[L:  7, P:  9]      |                        select_clause:
[L:  7, P:  9]      |                            keyword:                          'SELECT'
[L:  7, P: 15]      |                            [META] indent:
[L:  7, P: 15]      |                            whitespace:                       ' '
[L:  7, P: 16]      |                            select_clause_element:
[L:  7, P: 16]      |                                literal:                      '2'
[L:  7, P: 17]      |                        [META] dedent:
[L:  7, P: 17]      |                    newline:                                  '\n'
[L:  8, P:  1]      |                    whitespace:                               '    '
[L:  8, P:  5]      |                    [META] dedent:
[L:  8, P:  5]      |                    end_bracket:                              ')'
[L:  8, P:  6]      |            newline:                                          '\n'
[L:  9, P:  1]      |            newline:                                          '\n'
[L: 10, P:  1]      |            whitespace:                                       '    '
[L: 10, P:  5]      |            select_statement:
[L: 10, P:  5]      |                select_clause:
[L: 10, P:  5]      |                    keyword:                                  'SELECT'
[L: 10, P: 11]      |                    [META] indent:
[L: 10, P: 11]      |                    newline:                                  '\n'
[L: 11, P:  1]      |                    whitespace:                               '        '
[L: 11, P:  9]      |                    select_clause_element:
[L: 11, P:  9]      |                        column_reference:
[L: 11, P:  9]      |                            identifier:                       'cte_1'
[L: 11, P: 14]      |                            dot:                              '.'
[L: 11, P: 15]      |                            identifier:                       'var'
[L: 11, P: 18]      |                    comma:                                    ','
[L: 11, P: 19]      |                    newline:                                  '\n'
[L: 12, P:  1]      |                    whitespace:                               '        '
[L: 12, P:  9]      |                    select_clause_element:
[L: 12, P:  9]      |                        column_reference:
[L: 12, P:  9]      |                            identifier:                       'cte_2'
[L: 12, P: 14]      |                            dot:                              '.'
[L: 12, P: 15]      |                            identifier:                       'var'
[L: 12, P: 18]      |                newline:                                      '\n'
[L: 13, P:  1]      |                whitespace:                                   '    '
[L: 13, P:  5]      |                [META] dedent:
[L: 13, P:  5]      |                from_clause:
[L: 13, P:  5]      |                    keyword:                                  'FROM'
[L: 13, P:  9]      |                    whitespace:                               ' '
[L: 13, P: 10]      |                    from_expression:
[L: 13, P: 10]      |                        [META] indent:
[L: 13, P: 10]      |                        from_expression_element:
[L: 13, P: 10]      |                            table_expression:
[L: 13, P: 10]      |                                table_reference:
[L: 13, P: 10]      |                                    identifier:               'cte_1'
[L: 13, P: 15]      |                        [META] dedent:
[L: 13, P: 15]      |                    comma:                                    ','
[L: 13, P: 16]      |                    whitespace:                               ' '
[L: 13, P: 17]      |                    from_expression:
[L: 13, P: 17]      |                        [META] indent:
[L: 13, P: 17]      |                        from_expression_element:
[L: 13, P: 17]      |                            table_expression:
[L: 13, P: 17]      |                                table_reference:
[L: 13, P: 17]      |                                    identifier:               'cte_2'
[L: 13, P: 22]      |                        [META] dedent:
[L: 13, P: 22]      |    statement_terminator:                                     ';'
[L: 13, P: 23]      |    newline:                                                  '\n'
```

So we COULD look for any `common_table_expression` plus an optional `comma` and just check the next symbol is `newline` + `newline`, and if not insert it.

Instead what it currently does is:

1. Look for the `common_table_expression` direct children:

```sql
    cte_1 AS (
        ...
    ),

    cte_2 (...) AS (
        ...
    )
```

2. Look for bracketed segments
3. Check if `AS` is the code immediately before the bracketed - if so that's the brackets we want to look at for newlines

Currently step 3 is incorrectly coded as "check if `AS` is the first keyword before the bracketed segment", which means for `cte_2 (...)`, which has no keyword before it, it keeps on backtracking all the way into `cte_1`'s select definition until it finds a keyword there (`AS` in this example - which is the bug as it's the wrong `AS`).

I've basically changed it to look for other code, and not just keywords. This will ALWAYS match now as the cte name at the very least will match as code, which will prevent it from backtracking into the previous cte's select definition. So I think my current change is sufficient and think we should merge this as is. Perhaps the `preceding_code` logic could be further enhanced to check it never backtracks past the start of that CTE?

Ultimately I do think the rewrite to actually use the proper parse tree is potentially a better way but, as I say, I'm afraid that's a bit beyond me. Plus there are other edge cases to think of there too (comments, meta data, indents...etc). But happy to scrap this branch if you have time to work on that? If not I think we should merge as is.

Reply:
I'll take a quick look this morning. Happy to go ahead and merge if you like, as my proposal would be a rework/rewrite either way.

Reply:
@tunetheweb: I have a draft PR up.

#3490

It fixes the L022 issue. Some other tests may be failing -- I'm waiting on the build to see.